Ignore ErrNotFound when delete LB resources

FengyunPan opened the pull request:
IsNotFound error is fine since that means the object is
deleted already, so let's check it before return error.

FengyunPan commented:
/release-note-none

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -140,9 +140,6 @@ func getFloatingIPByPortID(client *gophercloud.ServiceClient, portID string) (*f
 			return false, err
 		}
 		floatingIPList = append(floatingIPList, f...)
-		if len(floatingIPList) > 1 {
```
This means the loop might keep going, only to error out later (which is fine and I agree it's an error case not worth optimising for).

Perhaps the final return line (L146) could become:
```go
  return len(floatingIPList) <= 1, nil
```
And this would achieve the same early-exit-on-multiple-results but with far less visual overhead.

---

Since this loop is such a common pattern in the openstack provider, an alternative that I've been meaning to do for a while would be to replace the whole lot with some sort of generic code that takes an extraction function (like `floatingips.ExtractFloatingIPs`), does this generic "build a list, and abort if you get more than one result" logic, and just returns a single (interface{}, error).  Again, I wish golang had parameterised types :(

FengyunPan replied:
@anguslees Thank for your review very much.
My modify just prevent it from checking {len(floatingIPList) > 1} again, and I see your comment, I will check it in loop and don't check it at line 157.

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -285,6 +273,9 @@ func getListenersByLoadBalancerID(client *gophercloud.ServiceClient, id string)
 		return true, nil
 	})
 	if err != nil {
+		if isNotFound(err) {
```
What's the benefit of pushing this down here (and similar elsewhere), rather than have the caller use `isNotFound(...)`?   From experience I'm nervous about ensuring that all functions correctly convert Openstack not-found errors into `ErrNotFound` - we've certainly had places where this was missed in the past.

FengyunPan replied:
Yeah, I know it, when len(...) is 0, K8S will return  ErrNotFound and caller can't use isNotFound(...). So I think the better way is that calling k8s function with ErrNotFound and calling OpenStack function with isNotFound(...). Do you think so?

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -703,39 +687,41 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
-					continue
-				} else {
-					return nil, fmt.Errorf("Error getting address for node %s: %v", node.Name, err)
+		if !isNotFound(err) && members != nil {
```
You have changed this code to separately use both a literal `err == ErrNotFound` *and* `isNotFound(err)`.  You should pick one :)

I suggest leaving this to use `isNotFound()` and remove all the places above where you've pushed this logic into the `getFooByBar` functions - because it's going to be a lot more fragile to rely on everyone returning just the `ErrNotFound` value.

FengyunPan replied:
You are right, I will update it.

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -1102,9 +1088,13 @@ func (lbaas *LbaasV2) EnsureLoadBalancerDeleted(clusterName string, service *v1.
 		}
+		if err == ErrNotFound {
+			// Unknown error, retry later
+			continue
```
Will this function be retried if it does *not* return an error?  It's been a while since I read the calling code, but I thought we had to return an error in order to be retried ...

FengyunPan replied:
Sorry, this description is incorrect. I will update it.

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -703,39 +687,41 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
+			for _, node := range nodes {
+				addr, err := nodeAddressForLB(node)
+				if err != nil {
+					if err == ErrNoAddressFound {
```
I think this `ErrNotFound` -> `ErrNoAddressFound` and the additional "ignore if not found" check around this foreach-node loop are the behavioural changes in this PR, and the rest is refactoring.  Is that correct or did I miss any others?

FengyunPan replied:
The nodeAddressForLB() can return ErrNoAddressFound, not ErrNotFound.
I do not know if it is correct, I will check it later.

NickrenREN commented:
@anguslees  @FengyunPan   Maybe we should keep consistent style to define Not Found Err.
Function return or caller check ?

FengyunPan commented:
@NickrenREN Ok.

NickrenREN commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -314,25 +298,18 @@ func getPoolByListenerID(client *gophercloud.ServiceClient, loadbalancerID strin
 	if len(listenerPools) == 0 {
 		return nil, ErrNotFound
-	} else if len(listenerPools) > 1 {
-		return nil, ErrMultipleResults
```
if len(listenerPools) >1 , do not return err ?

FengyunPan replied:
ListenerID is unique, can't get { len(listenerPools) >1}.
The getPoolByListenerID() just return Found or notFound, the check of 'if len(listenerPools) > 1' is useless.

NickrenREN replied:
Code showed that if len(listenerPools) > 1 , ErrMultipleResults was returned before, but after your change, we will never return ErrMultipleResults

anguslees replied:
I agree the check should never fire, but I'm nervous about trusting the OpenStack API to always do what we expect :)

On the other hand, I agree these additional checks add visual noise - so I only have a weak preference for keeping them.

FengyunPan replied:
Ok, keep them better.

FengyunPan replied:
Done

NickrenREN commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -703,46 +682,48 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
-					continue
-				} else {
-					return nil, fmt.Errorf("Error getting address for node %s: %v", node.Name, err)
+		if !lbaas.IsNotFound(err) && members != nil {
```
Why this ?  if members != nil and of course err returned is nil and do we really need this?

FengyunPan replied:
You are right, I delete it soon.

NickrenREN commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -1203,10 +1188,26 @@ func (lbaas *LbaasV2) EnsureLoadBalancerDeleted(clusterName string, service *v1.
 	return nil
 }
 
+func (lbaas *LbaasV2) IsNotFound(err error) bool {
```
Is this needed (and so is the similar function below)?

FengyunPan replied:
Yes, caller don't know getxxxx() return ErrNotFound or isNotFound err, we can use IsNotFound() to check err.

anguslees replied:
Just move this into the real `isNotFound()` - in fact I thought I had already extended that to check `ErrNotFound` some time ago, but apparently not...

FengyunPan replied:
Done

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -346,13 +323,19 @@ func getMembersByPoolID(client *gophercloud.ServiceClient, id string) ([]v2pools
 		}
 		members = append(members, membersList...)
-
+		if len(members) > 1 {
```
This particular function is intended to return multiple results - see pluralised name and array return type.

FengyunPan replied:
Done

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -288,6 +268,10 @@ func getListenersByLoadBalancerID(client *gophercloud.ServiceClient, id string)
 		return nil, err
 	}
 
+	if len(existingListeners) == 0 {
```
This function is intended to return a (possibly empty) list of results - ie: I think it's up to the caller to decide if len() == 0 is an error.

FengyunPan replied:
Done

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -346,13 +323,19 @@ func getMembersByPoolID(client *gophercloud.ServiceClient, id string) ([]v2pools
 		return nil, err
 	}
 
+	if len(members) == 0 {
```
Ditto - anything declared as returning a list is fine to return an empty list.

FengyunPan replied:
Done

anguslees commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -406,13 +388,19 @@ func getSecurityGroupRules(client *gophercloud.ServiceClient, opts rules.ListOpt
 			return false, err
 		}
 		securityRules = append(securityRules, ruleList...)
+		if len(securityRules) > 1 {
```
Same list return comments as above

FengyunPan replied:
Done

anguslees reviewed (commented):
(Thanks for your persistence @FengyunPan :)

Commit:
Ignore ErrNotFound when delete LB resources

IsNotFound error is fine since that means the object is
deleted already, so let's check it before return error.

FengyunPan commented:
@ anguslees Hi, I keep them and just ignore ErrNotFound when deleting LB resources.

dims commented:
@k8s-bot ok to test

dims commented:
/assign

dims commented:
/lgtm

dims commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

fejta commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this
ref https://github.com/kubernetes/test-infra/pull/2931

fejta commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

FengyunPan commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

NickrenREN commented:
@k8s-bot pull-kubernetes-kubemark-e2e-gce test this
@k8s-bot pull-kubernetes-federation-e2e-gce test this

NickrenREN commented:
/approve

FengyunPan commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

wojtek-t commented:
/approve no issue

wojtek-t commented:
/approve no-issue

FengyunPan commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

FengyunPan commented:
/test pull-kubernetes-federation-e2e-gce
